txpipe/twopoint_null_tests.py

```diff
@@ -766,3 +767,272 @@ def write_output(self, source_list, lens_list, meta, results):
+                      }
+    
+    def run(self):
+        import os
+        import pymaster as nmt
+        import healpy as hp
+        import sacc
+        import pyccl
+        import hybrideb
+        import datetime
+        import pickle
+        from tqdm import tqdm
+
+        if self.config["method"] == 'namaster_pureB' or self.config["method"] == 'namaster_nopureB':
+            '''B-mode method that is already implemented in NaMaster'''
+            print('----------------namaster-----------')
+            if  self.config["method"] == 'namaster_pureB':
+                print("WARNING: Namaster's B-mode purification requires the mask to be heavily apodized.")
+                print("For a realistic LSS mask, this will most likely not work, but this function is ")
+                print("implemented for completeness nonetheless.")
+                purify_b = True
+            else:
+                print("WARNING: We are not running B-mode purification (this is what was used in DES-Y3).")
+                purify_b = False
+                
+            # Open source maps (g1,g2,weights)
+            with self.open_input("source_maps", wrapper=True) as f:
+                nbin_source = f.file["maps"].attrs["nbin_source"]
+                g1_maps     = [f.read_map(f"g1_{b}") for b in range(nbin_source)]
+                g2_maps     = [f.read_map(f"g2_{b}") for b in range(nbin_source)]
+                weights     = [f.read_map(f"lensing_weight_{b}") for b in range(nbin_source)]
+
+            # Open mask
+            with self.open_input("mask", wrapper=True) as f:
+                mask = f.read_map("mask")
+            
+            # Define binning 
+            b     = nmt.NmtBin.from_nside_linear(hp.npix2nside(mask.shape[0]), self.config["Nell"])
+            
+            # Initialize the fields
+            fields = {}
+            for i in range(nbin_source):
+                g1_maps[i][g1_maps[i]==hp.UNSEEN]=0
+                g2_maps[i][g2_maps[i]==hp.UNSEEN]=0
+                fields[i] = nmt.NmtField(mask, [g1_maps[i], g2_maps[i]], purify_e=False, purify_b=purify_b)
+
+            # Compute Cls and store them in dictionrary
+            tmp       = np.array([])
+            ret=np.zeros((nbin_source,nbin_source,len(b.get_effective_ells())))
+
+            for zi in range(nbin_source):
+                for zj in range(zi,nbin_source):
+                     
+                    field1 = fields[zi]
+                    field2 = fields[zj]
+                    w_yp  = nmt.NmtWorkspace()
+                    w_yp.compute_coupling_matrix(field1, field2, b)
+    
+                    cl_coupled   = nmt.compute_coupled_cell(field1,field2)
+                    cl_decoupled = w_yp.decouple_cell(cl_coupled)
+                    ret[zj,zi,:]= cl_decoupled[3]
+                    tmp = np.concatenate([tmp,cl_decoupled[3]])
+
+            
+            # Compute covariance in a somewhat adhoc way by shuffling the pixel values.
```

Comment:
Does this definitely work? Could also do the think where you randomly rotate shears.

Reply:
I think what we typically do is rotate the shears at the catalog level not after projecting the galaxies onto pixels.
We would like to do that ideally but that requires passing down the full catalog to this stage and projecting iteratively. For now I've added pixel level rotation.

Reply:
That's great, thanks!